base/src/main/java/proguard/optimize/inline/BaseLambdaInliner.java

```diff
@@ -0,0 +1,343 @@
+import proguard.classfile.instruction.visitor.AllInstructionVisitor;
+import proguard.classfile.instruction.visitor.InstructionCounter;
+import proguard.classfile.instruction.visitor.InstructionOpCodeFilter;
+import proguard.classfile.instruction.visitor.InstructionVisitor;
+import proguard.classfile.util.ClassReferenceInitializer;
+import proguard.classfile.util.ClassUtil;
+import proguard.classfile.util.InternalTypeEnumeration;
+import proguard.classfile.visitor.MemberVisitor;
+import proguard.evaluation.PartialEvaluator;
+import proguard.evaluation.TracedStack;
+import proguard.optimize.inline.lambda_locator.Lambda;
+import proguard.optimize.peephole.MethodInliner;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.lang.Math.max;
+
+public class BaseLambdaInliner implements MemberVisitor, InstructionVisitor, ConstantVisitor {
+    private final Clazz consumingClass;
+    private final Method consumingMethod;
+    private final AppView appView;
+    private final Lambda lambda;
+    private final CodeAttributeEditor codeAttributeEditor;
+    private final boolean isStatic;
+    private final int calledLambdaIndex;
+    private final int sizeAdjustedLamdaIndex;
+    private final PartialEvaluator partialEvaluator;
+    private Clazz interfaceClass;
+    private Clazz lambdaClass;
+    private Method lambdaInvokeMethod;
+    private String bridgeDescriptor;
+    private Method inlinedLambdaMethod;
+    private Method staticInvokeMethod;
+    private InterfaceMethodrefConstant referencedInterfaceConstant;
+    private final List<Integer> invokeMethodCallOffsets;
+
+    public BaseLambdaInliner(AppView appView, Clazz consumingClass, Method consumingMethod, int calledLambdaIndex, Lambda lambda) {
+        this.consumingClass = consumingClass;
+        this.consumingMethod = consumingMethod;
+        this.appView = appView;
+        this.lambda = lambda;
+        this.codeAttributeEditor = new CodeAttributeEditor();
+        this.isStatic = (consumingMethod.getAccessFlags() & AccessConstants.STATIC) != 0;
+        this.calledLambdaIndex = calledLambdaIndex + (isStatic ? 0 : 1);
+        this.sizeAdjustedLamdaIndex = ClassUtil.internalMethodVariableIndex(consumingMethod.getDescriptor(consumingClass), true, this.calledLambdaIndex);
+        this.partialEvaluator = new PartialEvaluator();
+        this.invokeMethodCallOffsets = new ArrayList<>();
+    }
+
+    /**
+     * Idea: Inline a lambda by just giving it a consuming method and class in which it has to inline the particular
+     * lambda,  a reference to the newly created function with the inlined lambda is returned, the caller can then
+     * replace the call instruction with a call to this newly created function. This way we don't need to think about
+     * how it is called exactly, it can be used as a higher abstraction in other places.
+     */
+    public Method inline() {
+        if (consumingMethod instanceof LibraryMethod)
```

Comment:
You should just have the constructor take a `ProgramClass` and `ProgramMethod` instead. Also this is against the code style (always uses braces)

Reply:
The instanceof will have to happen somewhere in the program because it will find call instructions and just use .referencedMethod on the methodrefConstant so it seemed better to make the lambda inliner itself decide that it cannot inline LibraryMethods instead of some other unrelated class deciding that. That's why I decided to put it here.